src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -249,123 +256,167 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
-                    List<KeyValueChange> changedKeyValuesCollection = null;
                     Dictionary<string, ConfigurationSetting> data = null;
+                    Dictionary<string, ConfigurationSetting> ffCollectionData = null;
+                    bool collectionsUpdated = false;
+                    bool ffCollectionUpdated = false;
                     bool refreshAll = false;
                     StringBuilder logInfoBuilder = new StringBuilder();
                     StringBuilder logDebugBuilder = new StringBuilder();
 
                     await ExecuteWithFailOverPolicyAsync(clients, async (client) =>
                         {
                             data = null;
-                            watchedSettings = null;
+                            ffCollectionData = null;
+                            watchedIndividualKvs = null;
+                            watchedKvEtags = null;
+                            watchedFfEtags = null;
                             keyValueChanges = new List<KeyValueChange>();
-                            changedKeyValuesCollection = null;
+                            collectionsUpdated = false;
+                            ffCollectionUpdated = false;
                             refreshAll = false;
                             Uri endpoint = _configClientManager.GetEndpointForClient(client);
                             logDebugBuilder.Clear();
                             logInfoBuilder.Clear();
 
-                            foreach (KeyValueWatcher changeWatcher in refreshableWatchers)
+                            if (!_options.RegisterAllEnabled)
                             {
-                                string watchedKey = changeWatcher.Key;
-                                string watchedLabel = changeWatcher.Label;
+                                refreshAll = await RefreshIndividualKvWatchers(
+                                    client,
+                                    keyValueChanges,
+                                    refreshableIndividualKvWatchers,
+                                    endpoint,
+                                    logDebugBuilder,
+                                    logInfoBuilder,
+                                    cancellationToken).ConfigureAwait(false);
+
+                                if (!refreshAll)
+                                {
+                                    // Get feature flag changes
+                                    watchedFfEtags = new Dictionary<KeyValueSelector, IEnumerable<MatchConditions>>(_watchedFfEtags);
 
-                                KeyValueIdentifier watchedKeyLabel = new KeyValueIdentifier(watchedKey, watchedLabel);
+                                    ffCollectionUpdated = await UpdateWatchedCollections(
+                                        refreshableFfWatchers.Select(watcher => new KeyValueSelector { KeyFilter = watcher.Key, LabelFilter = watcher.Label }),
```

Comment:
Why are new instances of KeyValueWatcher necessary?

Reply:
I wanted to reuse this method to check if the collections have changed for both key-values and feature flags, but for key-values we want to check all selected kvs, and for flags we're still respecting the refresh interval for each watcher because of how `UseFeatureFlags` sets the refresh interval each time it's called for any flags selected. For key-values we just pass in all key-value selectors since we check them all for `RegisterAll`, while for feature flags we need to convert the feature flag watchers with elapsed refresh intervals into selectors somehow to pass to the helper method.

Reply:
Discussed offline, silly misread on my end. KeyValueSelectors are being constructed from watchers here, rather than copied over.